app/main.py

```diff
@@ -1,34 +1,96 @@
+from collections.abc import Generator
+from itertools import zip_longest
+from typing import Any
+from dataclasses import dataclass
+
+
+@dataclass
 class Deck:
-    def __init__(self, row, column, is_alive=True):
-        pass
+    row: int
+    column: int
+    is_alive: bool = True
 
 
 class Ship:
-    def __init__(self, start, end, is_drowned=False):
-        # Create decks and save them to a list `self.decks`
-        pass
+    def __init__(
+        self,
+        start: tuple[int, int],
+        end: tuple[int, int],
+        is_drowned: bool = False,
+    ) -> None:
+        self.start = start
+        self.end = end
+        self.is_drowned = is_drowned
+        self.decks: list[Deck] = [
+            Deck(i, j) for i, j in Battleship.generate_coordinates(start, end)
+        ]
 
-    def get_deck(self, row, column):
-        # Find the corresponding deck in the list
-        pass
+    def get_deck(self, row: int, column: int) -> Deck | None:
+        for deck in self.decks:
+            if (deck.row, deck.column) == (row, column):
+                return deck
+        return None
 
-    def fire(self, row, column):
-        # Change the `is_alive` status of the deck
-        # And update the `is_drowned` value if it's needed
-        pass
+    def fire(self, row: int, column: int) -> None:
+        deck: Deck | None = self.get_deck(row, column)
+        if deck is not None:
+            deck.is_alive = False
+            if not any(deck.is_alive for deck in self.decks):
+                self.is_drowned = True
 
 
 class Battleship:
-    def __init__(self, ships):
-        # Create a dict `self.field`.
-        # Its keys are tuples - the coordinates of the non-empty cells,
-        # A value for each cell is a reference to the ship
-        # which is located in it
-        pass
-
-    def fire(self, location: tuple):
-        # This function should check whether the location
-        # is a key in the `self.field`
-        # If it is, then it should check if this cell is the last alive
-        # in the ship or not.
-        pass
+    def __init__(
+        self, ships: list[tuple[tuple[int, int], tuple[int, int]]]
```

Comment:
I'd suggest using TypeVar here, because this typing might be correct, but I don't understand what is there

Reply:
could you, please, give me some hints on how to do that